### DISCUSSION TITLE
IfcOS C++ - internal data model documentation, design reasons?

### DISCUSSION THREAD

**Discussion by dirkolbrich** [IfcOpenShell/IfcOpenShell#4218] [General]
Hi there,

is there anywhere an in-depth documentation or write-up about the internal data model of the IFC hierarchy?

I am still trying to GROK this and would be interested to hear the reasoning behind the architectural decisions. I can see, that the parsed IFC file is stored in memory with a bunch of maps (some of the unordered hash maps) with the IfcFile class, but unfortunately I couldn't find the WHY behind this.

https://github.com/IfcOpenShell/IfcOpenShell/blob/b4120b1cab7de247a5ad9f9c959fc29c5db86c8c/src/ifcparse/IfcFile.h#L71-L84

**Comment by aothms** [comment]
Maybe this helps as a first starting point. Let me know which things need additional clarification.

	typedef std::map<const IfcParse::declaration*, aggregate_of_instance::ptr> entities_by_type_t;
    
Uses to query model based on IFC type    
    
	typedef boost::unordered_map<unsigned int, IfcUtil::IfcBaseClass*> entity_by_id_t;

Query by id; also instance references are resolved dynamically, parsing is only single pass    
   
	typedef boost::unordered_map<uint32_t, IfcUtil::IfcBaseClass*> entity_by_iden_t;

IfcBaseClass has a unique uint32_t in addition to the numeric instance from the step file. The  purpose for this is when you add an instance to a file we use the identity to see if that instance was already added once earlier and not add it again.

	typedef std::map<std::string, IfcUtil::IfcBaseClass*> entity_by_guid_t;
    
Query by guid
    
	typedef std::tuple<int, int, int> inverse_attr_record;
	enum INVERSE_ATTR { INSTANCE_ID, INSTANCE_TYPE, ATTRIBUTE_INDEX };
	typedef std::map<inverse_attr_record, std::vector<int> > entities_by_ref_t;
    
Used to resolve inverse attributes, which require the type and attribute index because inverse attributes have named forward attribute mirroring potentially a specific subtype.
    
	typedef std::map<int, std::vector<int> > entities_by_ref_excl_t;
    
Used for 'batch mode', which is basically to do a topological sort based for large subgraph deletion, so that you delete from the leafs inwards and don't keep updating inverses within the subgraph.
    
	typedef std::map<unsigned int, aggregate_of_instance::ptr> ref_map_t;
    
Appears unused. I think was once used for caching when inverse attribute calculation was more expensive.

**Comment by dirkolbrich** [comment]
Thank you for the overview.

So basically the parse functionality iterates over the file and creates for each # line (STEP instance) in the file a new `IfcEntityInstanceData` and adds a key and a pointer to this instance into the different maps if applicable.
But it initializes an `IfcEntityInstanceData` regardless if it is an IfcEntity, an IfcType, IfcFunction or IfcRule (as defined by the [schema]( https://standards.buildingsmart.org/IFC/RELEASE/IFC4_3/HTML/annex-b.html)) somewhere and doesn't keep them tightly packed in memory.

-> I think here is an opportunity for optimisation of the in-memory model. I'm thinking of using an Entity Component System known from game engines or reusing instances for IfcType e.g. do not duplicate an IfcCartesianPoint with the same value.

I had a hard time understanding the class declarations, either due to the change of naming style to suddenly snake_case compared to other elements and that class definitions are not really located in the files I would expect.

I also was a bit thrown off by the name `aggregate_of_instance` (it’s a term which is not commonly used in my daily language) but I found an older commit where it was named IfcEntityList, which made a lot more sense to me. I can guess why you did renamed it, maybe to more closely follow the [Aggregation concept ](https://standards.buildingsmart.org/IFC/RELEASE/IFC4_3/HTML/concepts/Object_Composition/Aggregation/content.html) of the IFC schema. Maybe I have to get used to it.

Here is a follow up question:

What’s the difference and purpose between the classes `aggregate_of_instance` and `aggregate_of_aggregate_of_instance` and their seemingly template sibling `aggregate_of` and `aggregate_of_aggregate_of`?

I get that first one as it basically is an Entity list / vector with added functionality.

**Comment by aothms** [comment]
> But it initializes an IfcEntityInstanceData regardless if it is an IfcEntity, an IfcType, IfcFunction or IfcRule 

Functions and Rules are not exchanged as part of the P21 (.stp) file format. For types it depends on the situation. In spf simple types are typically exchanged as the simple literal values they wrap. You don't read `IFCWALL(IFCGLOBALLYUNIQUEID('3w...'))` even though the first attribute of IfcWall is of type IfcGloballyUniqueId. Only in case of SELECT types such as the NominalValue in an IfcPropertySingleValue you will box the type in the serialization. And in those cases specifically an InstanceData with a single attribute is created (because that is what it looks like in the serialization).

> I also was a bit thrown off by the name aggregate_of_instance (it’s a term which is not commonly used in my daily language) but I found an older commit where it was named IfcEntityList, which made a lot more sense to me.

This is an unfortunate fact of life I think that if you want to remain true to the words in the specification you alienate the normal people. But entitylist is wrong for two reasons. It's not only applicable to lists in the schema, but also sets. Its values are not entities (which are the classes in STEP) but entity **instances**.

> What’s the difference and purpose between the classes aggregate_of_instance and aggregate_of_aggregate_of_instance and their seemingly template sibling aggregate_of and aggregate_of_aggregate_of?

This has to do with the separation of schema and parsed model. Only when you interact with a named attribute in the schema, the generically stored list of instances is casted to a typed (`aggregate_of ` have a template argument for the IFC entity class its values represent).

This is something we can do better now, to when parsing already use schema information to store values with less indirection based on the schema. Maybe IfcEntityInstanceData can actually be more like a tuple instead of an array with abstract values with virtual methods. But 10y ago, variadic templates where not there yet in C++ and in ifopsh we didn't have the runtime schema information available yet. This would be in my mind the most obvious memory and performance gain, there is really a lot of inefficiency in the virtual calls, casting (not only aggregate_of_instance  but also e.g vector<int>) and pointer chasing happening in the stored attribute values.
